File: tests/test_apps/elixir/test/smoke_test.exs
defmodule SmokeTest do
  use ExUnit.Case

  @fixtures_dir Path.join([__DIR__, "..", "..", "fixtures"])

  defp load_fixtures(name) do
    @fixtures_dir
    |> Path.join(name)
    |> File.read!()
    |> Jason.decode!()
  end

  describe "basic fixtures" do
    test "language_count is positive" do
      count = TreeSitterLanguagePack.language_count()
      assert count >= 100, "language_count #{count} < expected min 100"
    end

    test "has_language for known languages" do
      assert TreeSitterLanguagePack.has_language("python") == true
      assert TreeSitterLanguagePack.has_language("javascript") == true
      assert TreeSitterLanguagePack.has_language("rust") == true
      assert TreeSitterLanguagePack.has_language("go") == true
    end

    test "has_language returns false for nonexistent" do
      assert TreeSitterLanguagePack.has_language("nonexistent_xyz") == false
    end

    test "available_languages contains expected languages" do
      langs = TreeSitterLanguagePack.available_languages()
      assert is_list(langs)
      assert "python" in langs
      assert "javascript" in langs
      assert "rust" in langs
      assert "go" in langs
    end
  end

  describe "parse validation" do
    test "parses Python code" do
      {:ok, tree} = TreeSitterLanguagePack.parse_string("python", "def hello(): pass\n")
      {:ok, node_type} = TreeSitterLanguagePack.tree_root_node_type(tree)
      assert node_type == "module"

      {:ok, child_count} = TreeSitterLanguagePack.tree_root_child_count(tree)
      assert child_count >= 1

      {:ok, has_errors} = TreeSitterLanguagePack.tree_has_error_nodes(tree)
      assert has_errors == false
    end

    test "errors on invalid language" do
      assert {:error, _reason} = TreeSitterLanguagePack.parse_string("nonexistent_xyz_123", "code")
    end
  end

  describe "process tests" do
    setup do
      {:ok, fixtures} = File.read(Path.join(@fixtures_dir, "process.json"))
      fixtures_data = Jason.decode!(fixtures)
      {:ok, fixtures: fixtures_data}
    end

    test "python_function", %{fixtures: fixtures} do
      fixture = Enum.find(fixtures, fn f -> f["name"] == "python_function" end)
      assert fixture != nil

      config = Jason.encode!(fixture["config"])
      {:ok, result_json} = TreeSitterLanguagePack.process(fixture["source"], config)
      result = Jason.decode!(result_json)

      assert result["language"] == "python"
      assert length(result["structure"]) >= fixture["expected"]["structure_min"]
      assert result["metrics"]["error_count"] == 0
    end

    test "javascript_function", %{fixtures: fixtures} do
      fixture = Enum.find(fixtures, fn f -> f["name"] == "javascript_function" end)
      assert fixture != nil

      config = Jason.encode!(fixture["config"])
      {:ok, result_json} = TreeSitterLanguagePack.process(fixture["source"], config)
      result = Jason.decode!(result_json)

      assert result["language"] == "javascript"
      assert length(result["structure"]) >= fixture["expected"]["structure_min"]
    end

    test "rust_function", %{fixtures: fixtures} do
      fixture = Enum.find(fixtures, fn f -> f["name"] == "rust_function" end)
      assert fixture != nil

      config = Jason.encode!(fixture["config"])
      {:ok, result_json} = TreeSitterLanguagePack.process(fixture["source"], config)
      result = Jason.decode!(result_json)

      assert result["language"] == "rust"
      assert length(result["structure"]) >= fixture["expected"]["structure_min"]
    end

    test "python_with_imports", %{fixtures: fixtures} do
      fixture = Enum.find(fixtures, fn f -> f["name"] == "python_with_imports" end)
      assert fixture != nil

      config = Jason.encode!(fixture["config"])
      {:ok, result_json} = TreeSitterLanguagePack.process(fixture["source"], config)
      result = Jason.decode!(result_json)

      assert result["language"] == "python"
      assert length(result["imports"]) >= fixture["expected"]["imports_min"]
      assert length(result["structure"]) >= fixture["expected"]["structure_min"]
    end
  end

  describe "chunking tests" do
    setup do
      {:ok, fixtures} = File.read(Path.join(@fixtures_dir, "chunking.json"))
      fixtures_data = Jason.decode!(fixtures)
      {:ok, fixtures: fixtures_data}
    end

    test "python_chunking", %{fixtures: fixtures} do
      fixture = Enum.find(fixtures, fn f -> f["name"] == "python_chunking" end)
      assert fixture != nil

      config = Jason.encode!(fixture["config"])
      {:ok, result_json} = TreeSitterLanguagePack.process(fixture["source"], config)
      result = Jason.decode!(result_json)

      assert length(result["chunks"]) >= fixture["expected"]["chunks_min"]
    end

    test "javascript_chunking", %{fixtures: fixtures} do
      fixture = Enum.find(fixtures, fn f -> f["name"] == "javascript_chunking" end)
      assert fixture != nil

      config = Jason.encode!(fixture["config"])
      {:ok, result_json} = TreeSitterLanguagePack.process(fixture["source"], config)
      result = Jason.decode!(result_json)

      assert length(result["chunks"]) >= fixture["expected"]["chunks_min"]
    end
  end

  describe "download API" do
    test "downloaded_languages returns list" do
      {:ok, langs} = TreeSitterLanguagePack.downloaded_languages()
      assert is_list(langs)
      assert length(langs) > 0
    end

    test "cache_dir returns string" do
      {:ok, cache_dir} = TreeSitterLanguagePack.cache_dir()
      assert is_binary(cache_dir)
      assert byte_size(cache_dir) > 0
    end

    test "manifest_languages returns 170+ items" do
      {:ok, langs} = TreeSitterLanguagePack.manifest_languages()
      assert is_list(langs)
      assert length(langs) >= 170
    end

    test "process with invalid language errors" do
      config = Jason.encode!(%{"language" => "nonexistent_xyz"})
      assert {:error, _reason} = TreeSitterLanguagePack.process("code", config)
    end
  end

  describe "error handling" do
    test "has_language returns false for nonexistent" do
      assert TreeSitterLanguagePack.has_language("nonexistent_lang_xyz_123") == false
    end
  end

  describe "setup" do
    test "initializes with multiple languages" do
      config = Jason.encode!(%{"languages" => ["python", "javascript", "rust", "go", "ruby", "java", "c", "cpp"]})
      assert :ok = TreeSitterLanguagePack.init(config)
    end
  end
end

File: tests/test_apps/java/src/test/java/dev/kreuzberg/test/SmokeTest.java
package dev.kreuzberg.test;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import dev.kreuzberg.TreeSitterLanguagePack;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

class SmokeTest {

    private static final Gson GSON = new Gson();
    private static final Path FIXTURES_DIR = Path.of("..", "fixtures");

    @BeforeAll
    static void setup() {
        // Download required languages
        TreeSitterLanguagePack.download(List.of("python", "javascript", "rust", "go", "ruby", "java", "c", "cpp"));
    }

    @SuppressWarnings("unchecked")
    private static List<Map<String, Object>> loadFixtures(String name) throws IOException {
        String json = Files.readString(FIXTURES_DIR.resolve(name));
        Type type = new TypeToken<List<Map<String, Object>>>() {}.getType();
        return GSON.fromJson(json, type);
    }

    // Basic fixtures tests
    @TestFactory
    Stream<DynamicTest> basicFixtures() throws IOException {
        List<Map<String, Object>> fixtures = loadFixtures("basic.json");

        return fixtures.stream().map(fixture -> DynamicTest.dynamicTest(
            (String) fixture.get("name"),
            () -> {
                String test = (String) fixture.get("test");
                switch (test) {
                    case "language_count" -> {
                        int count = TreeSitterLanguagePack.languageCount();
                        int expectedMin = ((Number) fixture.get("expected_min")).intValue();
                        assertTrue(count >= expectedMin,
                            "language_count " + count + " < expected min " + expectedMin);
                    }
                    case "has_language" -> {
                        String language = (String) fixture.get("language");
                        boolean result = TreeSitterLanguagePack.hasLanguage(language);
                        boolean expected = (Boolean) fixture.get("expected");
                        assertEquals(expected, result,
                            "has_language(" + language + ") = " + result + ", expected " + expected);
                    }
                    case "available_languages" -> {
                        List<String> langs = TreeSitterLanguagePack.availableLanguages();
                        @SuppressWarnings("unchecked")
                        List<String> expectedContains = (List<String>) fixture.get("expected_contains");
                        for (String lang : expectedContains) {
                            assertTrue(langs.contains(lang),
                                "available_languages missing '" + lang + "'");
                        }
                    }
                    default -> fail("Unknown test type: " + test);
                }
            }
        ));
    }

    // Process fixtures tests
    @TestFactory
    Stream<DynamicTest> processFixtures() throws IOException {
        List<Map<String, Object>> fixtures = loadFixtures("process.json");

        return fixtures.stream().map(fixture -> DynamicTest.dynamicTest(
            (String) fixture.get("name"),
            () -> {
                String source = (String) fixture.get("source");
                @SuppressWarnings("unchecked")
                Map<String, Object> configMap = (Map<String, Object>) fixture.get("config");
                @SuppressWarnings("unchecked")
                Map<String, Object> expected = (Map<String, Object>) fixture.get("expected");

                String configJson = GSON.toJson(configMap);
                Map<String, Object> result = TreeSitterLanguagePack.process(source, configJson);

                if (expected.containsKey("language")) {
                    assertEquals(expected.get("language"), result.get("language"));
                }
                if (expected.containsKey("structure_min")) {
                    @SuppressWarnings("unchecked")
                    List<Object> structure = (List<Object>) result.get("structure");
                    int min = ((Number) expected.get("structure_min")).intValue();
                    assertTrue(structure.size() >= min,
                        "structure count " + structure.size() + " < min " + min);
                }
                if (expected.containsKey("imports_min")) {
                    @SuppressWarnings("unchecked")
                    List<Object> imports = (List<Object>) result.get("imports");
                    int min = ((Number) expected.get("imports_min")).intValue();
                    assertTrue(imports.size() >= min,
                        "imports count " + imports.size() + " < min " + min);
                }
                if (expected.containsKey("error_count")) {
                    @SuppressWarnings("unchecked")
                    Map<String, Object> metrics = (Map<String, Object>) result.get("metrics");
                    int errorCount = ((Number) metrics.get("error_count")).intValue();
                    int expectedCount = ((Number) expected.get("error_count")).intValue();
                    assertEquals(expectedCount, errorCount);
                }
                if (expected.containsKey("metrics_total_lines_min")) {
                    @SuppressWarnings("unchecked")
                    Map<String, Object> metrics = (Map<String, Object>) result.get("metrics");
                    int totalLines = ((Number) metrics.get("total_lines")).intValue();
                    int min = ((Number) expected.get("metrics_total_lines_min")).intValue();
                    assertTrue(totalLines >= min,
                        "total_lines " + totalLines + " < min " + min);
                }
            }
        ));
    }

    // Chunking fixtures tests
    @TestFactory
    Stream<DynamicTest> chunkingFixtures() throws IOException {
        List<Map<String, Object>> fixtures = loadFixtures("chunking.json");

        return fixtures.stream().map(fixture -> DynamicTest.dynamicTest(
            (String) fixture.get("name"),
            () -> {
                String source = (String) fixture.get("source");
                @SuppressWarnings("unchecked")
                Map<String, Object> configMap = (Map<String, Object>) fixture.get("config");
                @SuppressWarnings("unchecked")
                Map<String, Object> expected = (Map<String, Object>) fixture.get("expected");

                String configJson = GSON.toJson(configMap);
                Map<String, Object> result = TreeSitterLanguagePack.process(source, configJson);

                if (expected.containsKey("chunks_min")) {
                    @SuppressWarnings("unchecked")
                    List<Object> chunks = (List<Object>) result.get("chunks");
                    int min = ((Number) expected.get("chunks_min")).intValue();
                    assertTrue(chunks.size() >= min,
                        "chunks count " + chunks.size() + " < min " + min);
                }
            }
        ));
    }

    // Download API tests
    @Test
    void downloadedLanguagesReturnsArray() {
        List<String> langs = TreeSitterLanguagePack.downloadedLanguages();
        assertNotNull(langs);
        assertTrue(langs instanceof List);
    }

    @Test
    void manifestLanguagesReturnsArrayWith50Plus() {
        List<String> langs = TreeSitterLanguagePack.manifestLanguages();
        assertNotNull(langs);
        assertTrue(langs.size() > 50, "manifestLanguages should return 50+ languages");
    }

    @Test
    void cacheDirReturnsNonEmptyString() {
        String dir = TreeSitterLanguagePack.cacheDir();
        assertNotNull(dir);
        assertTrue(dir.length() > 0, "cacheDir should return non-empty string");
    }

    @Test
    void initDoesNotThrow() {
        assertDoesNotThrow(() -> TreeSitterLanguagePack.init());
    }

    // Parse validation tests
    @Test
    void parsesPythonCode() {
        var tree = TreeSitterLanguagePack.parseString("python", "def hello(): pass\n");
        assertNotNull(tree);
        assertEquals("module", tree.rootNodeType());
        assertTrue(tree.rootChildCount() >= 1);
        assertFalse(tree.hasErrorNodes());
    }

    @Test
    void errorsOnInvalidLanguage() {
        assertThrows(Exception.class, () ->
            TreeSitterLanguagePack.parseString("nonexistent_xyz_123", "code"));
    }

    @Test
    void hasLanguageReturnsFalseForNonexistent() {
        boolean result = TreeSitterLanguagePack.hasLanguage("nonexistent_xyz_123");
        assertFalse(result);
    }
}
